### PR TITLE
Add verify-redirects command

### LINKED ISSUE #13: Add Verification CLI method
Use Case:
- Run a CSV or other bulk import
- We now want to make sure that the redirects work as expected

It would be interesting to be able to run a cli command (or something similar) that verifies that the redirects were pulled in successfully and work as expected.

### PR DESCRIPTION
**Pull request by rclachance**
Resolves #13

This command verifies the redirects are working properly, and marks them as published (to denote they have been verified).

### What we're checking for (in plain english)

1. If redirecting to a URL:
	- Does the URL pass `wp_validate_redirect()`, which checks for a valid url and an allowed host to redirect to.

2. If redirecting to a post id:
	- Does the post we're redirecting to exist?
	- Is the post we're redirecting to published?
	- Are we redirecting to a post type that is not publicly accessible?

3. If we haven't failed validation yet:
	- Test the redirect

	- Is the redirect a 300 level status code?
		- If the redirect matches our destination URL, then mark the redirect as published (verified)
		- Did we redirect to the same page, but with a trailing slash?
		- If we reach this point, the redirect failed.

	- Is the redirect a 200 level status code?
		- If we reach this point, the redirect never ran. The from URL must not be a 404.

	- Is it a different status code? (not 200 or 300 level)
		- Something bad happened, report the status code.

### Performance features & notes

* We're querying for 500 redirect entries at a time, sleeping for 1 second every 100 rows processed.
* The main query uses a left join on the same table to capture post parent info that allows us to run all the checks in section 2 without calling `get_post()`. 
This allows us to fail early if there's a problem before we need to call `get_post()` (the most expensive query we're using).
* The query fetches rows ordered by post_parent, so multiple redirects to the same post are verified one after another. I've added code that will reuse the post object in scenarios like this, cutting down on unnecessary calls.

The most expensive verification will be redirects to post ids that pass verification and need to be marked as published (verified). Failure notices should be optimized to happen as quickly and efficiently as possible.

### Testing Instructions

* Setup a blank site on your localhost of choice
* run the following commands in terminal:
	* `wp plugin install wordpress-importer --activate`
	* `wp import --prompt`
		* download a copy of the file from https://wpcom-themes.svn.automattic.com/demo/theme-unit-test-data.xml
		* --author=create
* Using the following codeblock, make a csv file and import it into the site:
```
/redirect-to-non-url,this-isnt-a-url
/redirect-to-a-bad-url,/lot$-0f-+_)(*&^%$#!—Junk
/redirect-to-a-non-whitelisted-host,https://google.com
/redirect-to-autodraft,3
/redirect-to-scheduled-post,1153
/redirect-to-attachment,1692
/redirect-to-published-post,993
/redirect-to-published-page,735
notaurl-redirecting-topublishedpage/735
/lot$-0f-+_)(*&^%$#!—Junk/735
/post-id-that-doesnt-exist/20000
/sample-page/,/anywhere
/sample-page,this-should-redirect-to-include-trailing-slash
```
* in terminal, run `wp wpcom-legacy-redirector verify-redirects --format=table`

### CODE REVIEW

**Comment by rclations** [comment]
cc @mjangda

File: wpcom-legacy-redirector.php [hunk 1]
```diff
@@ -170,12 +171,112 @@ static function get_redirect_post_id( $url ) {
...
+
+	/**
+	 * Verify a redirect.
+	 *
+	 * @param array $redirect Redirect array.
+	 * @return bool|string True if the redirect passes validation. Array containing error notice if redirect fails validation.
+	 */
+	static function verify_redirect( $redirect ) {
+
+		if ( 3 == substr( $redirect['redirect']['status'], 0, 1 ) ) {
```
Comment (mjangda) [inline wpcom-legacy-redirector.php]: This feels like an overly complicated way to verify the status code. We should either add an inline comment explaining what we're trying to do, or (preferably) just simplify to a greater than/less than check instead.

File: wpcom-legacy-redirector.php [hunk 2]
```diff
@@ -170,12 +171,112 @@ static function get_redirect_post_id( $url ) {
...
+				'message'   => 'Did not redirect - returned ' . $redirect['redirect']['status'],
+			);
+		} else {
+			return array(
+				'id'        => $redirect['id'],
+				'from_url'  => $redirect['from']['path'],
+				'to_url'    => $redirect['to']['raw'],
+				'message'   => $redirect['redirect']['status'] . ' ' . $redirect['redirect']['status_msg'],
+			);
+		} // End if().
```
Comment (mjangda) [inline wpcom-legacy-redirector.php]: Minor: unnecessary inline comment.

File: wpcom-legacy-redirector.php [hunk 3]
```diff
@@ -170,12 +171,112 @@ static function get_redirect_post_id( $url ) {
...
+				);
+			} else {
+				return array(
+					'id'        => $redirect['id'],
+					'from_url'  => $redirect['from']['path'],
+					'to_url'    => $redirect['to']['raw'],
+					'message'   => 'Mismatch: redirected to ' . $redirect['redirect']['resulting_url'],
+				);
+			}
+		} elseif ( 2 == substr( $redirect['redirect']['status'], 0, 1 ) ) {
```
Comment (mjangda) [inline wpcom-legacy-redirector.php]: Same note as above about simplifying the check.

File: wpcom-legacy-redirector.php [hunk 4]
```diff
@@ -170,12 +171,112 @@ static function get_redirect_post_id( $url ) {
...
+	 * @return bool|string True if the redirect passes validation. Array containing error notice if redirect fails validation.
+	 */
+	static function verify_redirect( $redirect ) {
+
+		if ( 3 == substr( $redirect['redirect']['status'], 0, 1 ) ) {
+			if ( $redirect['to']['formatted'] === $redirect['redirect']['resulting_url'] ) {
+				return true;
+			} elseif ( $redirect['redirect']['resulting_url'] === $redirect['from']['url'] . '/' ) {
+				return array(
+					'id'        => $redirect['id'],
```
Comment (mjangda) [inline wpcom-legacy-redirector.php:389]: Most of the outputs across the various conditions seems to be repeated; should we reduce the repetition a bit?

File: wpcom-legacy-redirector.php [hunk 5]
```diff
@@ -170,12 +171,112 @@ static function get_redirect_post_id( $url ) {
...
+		} // End if().
+
+		return true;
+	}
+
+	/**
+	 * Verify a redirect.
+	 *
+	 * @param array $redirect Redirect array.
+	 * @return bool|string True if the redirect passes validation. Array containing error notice if redirect fails validation.
```
Comment (mjangda) [inline wpcom-legacy-redirector.php]: The return type is incorrect here.

Also, any reason to return an array instead of WP_Error?

File: wpcom-legacy-redirector.php [hunk 6]
```diff
@@ -170,12 +171,112 @@ static function get_redirect_post_id( $url ) {
...
+
+		if ( 3 == substr( $redirect['redirect']['status'], 0, 1 ) ) {
+			if ( $redirect['to']['formatted'] === $redirect['redirect']['resulting_url'] ) {
+				return true;
+			} elseif ( $redirect['redirect']['resulting_url'] === $redirect['from']['url'] . '/' ) {
+				return array(
+					'id'        => $redirect['id'],
+					'from_url'  => $redirect['from']['path'],
+					'to_url'    => $redirect['to']['raw'],
+					'message'   => 'Did not redirect to new page (only to add trailing slash).',
```
Comment (mjangda) [inline wpcom-legacy-redirector.php]: This message seems a bit unclear.

File: wpcom-legacy-redirector.php [hunk 7]
```diff
@@ -170,12 +171,112 @@ static function get_redirect_post_id( $url ) {
...
+				);
+			}
+		} elseif ( 2 == substr( $redirect['redirect']['status'], 0, 1 ) ) {
+			return array(
+				'id'        => $redirect['id'],
+				'from_url'  => $redirect['from']['path'],
+				'to_url'    => $redirect['to']['raw'],
+				'message'   => 'Did not redirect - returned ' . $redirect['redirect']['status'],
+			);
+		} else {
```
Comment (mjangda) [inline wpcom-legacy-redirector.php:431]: Minor: We can remove this last `else`. Not really needed.

File: wpcom-legacy-redirector.php [hunk 8]
```diff
@@ -170,12 +171,112 @@ static function get_redirect_post_id( $url ) {
...
+
+			if ( ! in_array( $redirect['parent']['post_type'], $post_types ) ) {
+				return array(
+					'id'        => $redirect['id'],
+					'from_url'  => $redirect['from']['path'],
+					'to_url'    => $redirect['to']['raw'],
+					'message'   => 'Attempting to redirect to a private post type: ' . $redirect['parent']['post_type'],
+				);
+			}
+		} // End if().
```
Comment (mjangda) [inline wpcom-legacy-redirector.php]: Minor: unnecessary inline comment.

File: wpcom-legacy-redirector.php [hunk 9]
```diff
@@ -170,12 +171,112 @@ static function get_redirect_post_id( $url ) {
...
+					'id'        => $redirect['id'],
+					'from_url'  => $redirect['from']['path'],
+					'to_url'    => $redirect['to']['formatted'],
+					'message'   => 'failed wp_validate_redirect()',
+				);
+			}
+
+		} else {
+
+			if ( ! $redirect['parent']['id'] > 0 ) {
```
Comment (mjangda) [inline wpcom-legacy-redirector.php]: Why not `$redirect['parent']['id'] <= 0` or `empty( $redirect['parent']['id'] )`

File: wpcom-legacy-redirector.php [hunk 10]
```diff
@@ -170,12 +171,112 @@ static function get_redirect_post_id( $url ) {
...
 
 		return $redirect_post_id;
 	}
 
+	/**
+	 * Validate the redirect should work normally.
+	 *
+	 * @param array $redirect Redirect array.
+	 * @param array $post_types Array of publicly accessible post types.
+	 * @return bool|string True if the redirect passes validation. Array containing error notice if redirect fails validation.
```
Comment (mjangda) [inline wpcom-legacy-redirector.php]: Same notes as verify_redirect around return type being incorrect and returning a WP_Error.

File: wpcom-legacy-redirector.php [hunk 11]
```diff
@@ -170,12 +171,112 @@ static function get_redirect_post_id( $url ) {
...
+	 */
+	static function validate_redirect( $redirect, $post_types ) {
```
Comment (mjangda) [inline wpcom-legacy-redirector.php]: Should we separate this out into separate methods: one for url and one for post?

File: includes/wp-cli.php [hunk 12]
```diff
@@ -163,15 +165,252 @@ function import_from_csv( $args, $assoc_args ) {
...
+						$update_redirect_status['publish'][] = $redirect['id'];
+					}
+				} else {
+					$notices[] = $verify;
+					if ( 'draft' !== $redirect['post_status'] ) {
+						$update_redirect_status['draft'][] = $redirect['id'];
+					}
+					continue;
+				}
+			} // End foreach().
```
Comment (mjangda) [inline includes/wp-cli.php]: Minor: unnecessary inline comment.

File: includes/wp-cli.php [hunk 13]
```diff
@@ -163,15 +165,252 @@ function import_from_csv( $args, $assoc_args ) {
...
+
+			// Update redirect status
+			if ( count( $update_redirect_status ) > 0 ) {
+				foreach ( $update_redirect_status as $redirect_status => $redirects_to_update ) {
+					foreach ( $redirects_to_update as $redirect_to_update ) {
+						$updated_rows = $wpdb->update( $wpdb->posts, array( 'post_status' => $redirect_status ), array( 'ID' => $redirect_to_update ) );
```
Comment (mjangda) [inline includes/wp-cli.php]: We'll need to trigger a `clean_post_cache` for these posts otherwise we'll end up have stale values that are mismatched.

File: includes/wp-cli.php [hunk 14]
```diff
@@ -163,15 +165,252 @@ function import_from_csv( $args, $assoc_args ) {
...
+	 * : Enable verification for validated redirects to post ids.
+	 *
+	 * ## EXAMPLES
+	 *
+	 * wp wpcom-legacy-redirector verify-redirects
+	 *
+	 * @subcommand verify-redirects
+	 * @synopsis [--status=<status>] [--format=<format>] [--verbose] [--strict]
+	 */
+	function verify_redirects( $args, $assoc_args ) {
```
Comment (mjangda) [inline includes/wp-cli.php:218]: The code for this command is starting to feel really long and I'm wondering if we should break it down into smaller methods?

File: includes/wp-cli.php [hunk 15]
```diff
@@ -163,15 +165,252 @@ function import_from_csv( $args, $assoc_args ) {
...
+				}
+
+				// Ping the URL and get the headers
+				$redirect_head = wp_remote_head( $redirect['from']['url'] );
+				if ( is_wp_error( $redirect_head ) ) {
+					$notices[] = array(
+						'id'        => $redirect['id'],
+						'from_url'  => $redirect['from']['path'],
+						'to_url'    => $redirect['to']['raw'],
+						'message'   => implode( "\n", $redirect_head->get_error_messages() ),
```
Comment (mjangda) [inline includes/wp-cli.php]: Minor: PHP_EOL.

File: includes/wp-cli.php [hunk 16]
```diff
@@ -163,15 +165,252 @@ function import_from_csv( $args, $assoc_args ) {
...
+	 * @synopsis [--status=<status>] [--format=<format>] [--verbose] [--strict]
+	 */
+	function verify_redirects( $args, $assoc_args ) {
+
+		global $wpdb;
+		$post_types = get_post_types( array( 'public' => true ) );
+
+		$status = \WP_CLI\Utils\get_flag_value( $assoc_args, 'status' );
+		switch ( $status ) {
+			case 'unverified':
```
Comment (mjangda) [inline includes/wp-cli.php:236]: I'm not clear on what the expected use for the verification flag is. Is it to avoid activating known broken redirects or something else?

**Review by mjangda — CHANGES_REQUESTED** [review]
Thanks for working through this. i've left a bunch of notes, questions, and areas for improvement.

One big thing that we should look at adding are tests (which I know you referenced in our discussion and would be nice to see hear alongside the newly added code).

File: includes/wp-cli.php (hunk 16 above)
Comment (rclachance) [reply includes/wp-cli.php:236]: The default is to only process redirects that have not previously passed verification (unverified). The status flag allows this to be overridden.

- The default behavior will save a lot of time in the event of multiple batches of redirects.
- If the permalink structure changed on the site for any reason, re-verifying all redirects could be necessary.
- I don't see a huge use case for only reverifying previously verified URLs, but it seemed logical to include as an option here and required minimal overhead.

Comment (rclachance) [reply includes/wp-cli.php:236]: Your mention of "activating known broken redirects" brings up an interesting point. Verification status doesn't currently have any effect over whether a redirect is used, but that could be something interesting to explore - especially if we moved closer to validation vs. verification.

File: includes/wp-cli.php (hunk 14 above)
Comment (rclachance) [reply includes/wp-cli.php:218]: my last commit moved the validation & verification processes into their own functions, leaving the main loops (which are tied to the cli progressmeter) in the main command.

do you have any suggestions for if/how this could be further broken down into separate methods?

File: wpcom-legacy-redirector.php (hunk 7 above)
Comment (rclachance) [reply wpcom-legacy-redirector.php:431]: I think we'll want this for reporting 400 and 500-level status codes?

**Comment by rclations** [comment]
I've restructured a lot of the code here and switched our verification method to parallel processing via curl to strike a better balance between speed & performance.

PHP 7.0.7 or higher is required to properly limit concurrent connections, so I've added some manual limits that slow things way down in lower php versions.

The `[--force_ssl]` flag is new, and allows for verifying relative URLs over SSL (to be used when the site can be accessed over SSL, but is not SSL by default). The benefit here is that over SSL we can use HTTP/2 multiplexing, which offers true concurrent processing over HTTP/1 pipelining (which process sequentially, but reuses an existing connection to do so).

Ran some tests using a free Pantheon development site to get some performance benchmarks (using > PHP 7.07). NewRelic screenshot of those tests w/ additional info is below:

---
NewRelic Screenshot

1. `wp wpcom-legacy-redirector import-from-csv`
	- 3.12% of CPU max

2. `wp wpcom-legacy-redirector verify-redirects -- --verbose --format=table`
	- max 20 concurrent connections
	- ~ 9s execution time
	- 24.7% of CPU max

3. `wp wpcom-legacy-redirector verify-redirects -- --verbose --format=table --status=all --force_ssl`
	- max 20 concurrent connections
	- ~ 19s execution time
	- 50.6% of CPU max

4. `wp wpcom-legacy-redirector verify-redirects -- --format=table --status=all`
	- max 10 concurrent connections (7f1f2ad)
	- ~ 16s execution time
	- 46.6% of CPU max

5. `wp wpcom-legacy-redirector verify-redirects -- --format=table --status=all --force_ssl`
	- max 10 concurrent connections (7f1f2ad)
	- ~ 15s execution time
	- 45.3% of CPU max

**Comment by GaryJones** [comment]
Thank you for this, @rclachance, and for the care that went into it — the parallel curl work and the performance benchmarking were well ahead of where the plugin was at the time.

The codebase this patches was replaced wholesale in #159, so the branch can no longer be merged as-is. Much of what you set out to do is now in 2.0.0 as `wp wpcom-legacy-redirector validate` ([`ValidateCommand.php`](https://github.com/Automattic/wpcom-legacy-redirector/blob/develop/src/Infrastructure/WordPress/Cli/ValidateCommand.php), [`RedirectValidator.php`](https://github.com/Automattic/wpcom-legacy-redirector/blob/develop/src/Application/RedirectValidator.php)): destination checks for deleted, trashed and unpublished posts, optional HTTP checks on destination URLs, `--status`, `--format`, and `--fix` to disable anything broken. There is also a single-redirect mode this PR did not have.

The part we have not carried over is the one I think was most distinctive here: actually requesting each source URL and confirming both that the redirect fires and that it lands on the intended destination. That is still a good idea, and I have opened #206 to track it as an opt-in flag on `validate`, referencing this PR for the design and for the failure classes you defined (did-not-redirect, mismatch, hop count).

Closing as superseded, with thanks.
